functional_programming_tests/test/main.cpp

```diff
@@ -0,0 +1,250 @@
+    Map<unsigned char> map;
+    // Get the costmap from a ros topic
+    // Check that map layout makes sense
+    if ((costmap.layout.dim[0].size * costmap.layout.dim[1].size) !=
+        costmap.layout.dim[0].stride) {
+      return std::nullopt;
+    }
+    if (costmap.layout.dim[0].stride != costmap.data.size()) {
+      return std::nullopt;
+    }
+    auto const begin = std::begin(costmap.data);
+    // Populate the map
+    for (size_t row = 0; row < costmap.layout.dim[0].size; row++) {
+      map.get_data().push_back(
+          {begin + row * costmap.layout.dim[1].size,
+           begin + (row + 1) * costmap.layout.dim[1].size});
+    }
+    return map;    
+  }
+
+  std_msgs::msg::UInt8MultiArray createUInt8MultiArrayMessageFromPath(const Path& path) {
+    auto message = std_msgs::msg::UInt8MultiArray();
+
+    message.layout.dim.resize(3, std_msgs::msg::MultiArrayDimension());
+
+    message.layout.dim[0].label = "rows";
+    message.layout.dim[0].size = path.size();
+    message.layout.dim[0].stride = path.size() * 2;
+
+    message.layout.dim[1].label = "columns";
+    message.layout.dim[1].size = 2;
+    message.layout.dim[1].stride = 1;
+
+    message.layout.dim[2].label = "channel";
+    message.layout.dim[2].size = 1;
+    message.layout.dim[2].stride = 1;
+
+    // Start pushing back the path only if there is one
+    if (path.size() > 0) {
+      for (auto const& position : path) {
+        message.data.push_back(position.x);
+        message.data.push_back(position.y);
+      }
+    }
+    return message;
+  }
+
+};
+
+Map<unsigned char> set_costmap_callback(const std::shared_ptr<example_srvs::srv::SetMap::Request> request,
+                                        std::shared_ptr<example_srvs::srv::SetMap::Response> response) {
+  response->success.data = false;
+  Map<unsigned char> costmap;
+  const auto parsed_map = PathGeneratorUtilities::parseSetMapRequest(request);
+  if (parsed_map.has_value()) {
+    response->success.data = true;
+    costmap = parsed_map.value();
+  }
+  return costmap;
+}
+
+// TODO: The return of this function can be the response
+// TODO: Edit this function such that the generate_path function is an input
+// TODO: Set the return  to be a std::expected with the various errors ()
+// TODO: Make a function that takes the error
+std::shared_ptr<example_srvs::srv::GetPath::Response> generate_path_callback(
+                           const std::shared_ptr<example_srvs::srv::GetPath::Request> request,
+                           const Map<unsigned char>& costmap,
+                           std::function<std::optional<Path>(Position const&, Position const&, Map<unsigned char> const&)> path_generater_function)
+{
+  std::shared_ptr<example_srvs::srv::GetPath::Response> response = std::make_shared<example_srvs::srv::GetPath::Response>();
+  if (costmap.get_data().size() == 0) {
+    response->success.data = false;
+    response->path = std_msgs::msg::UInt8MultiArray();
+    return response;
```

Comment:
Does
```suggestion
  if (costmap.get_data().size() == 0) {
     return {false, std_msgs::msg::UInt8MultiArray()};
```
or some variant like
```suggestion
  if (costmap.get_data().size() == 0) {
     return {success.data=false, path=std_msgs::msg::UInt8MultiArray()};
```
or
```suggestion
  if (costmap.get_data().size() == 0) {
     return std::make_shared<example_srvs::srv::GetPath::Response>{success.data=false, path=std_msgs::msg::UInt8MultiArray()};
```
work?
Wondering in case we can avoid/delay creating the response object for as long as possible.

Reply:
None of those worked, but I was able to replace them with an empty response lambda that creates the return object when it's called?